### PR TITLE
feat(argo-server): Add support for SSO. See #1813

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to USERS.md.

See https://github.com/alexec/argo/pull/3

# Screenshots

Screen Shot 2020-05-15 at 15 50 05

### CODE REVIEW

File: manifests/quick-start/base/overlays/argo-server-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  name: argo-server
+spec:
+  template:
+    spec:
+      containers:
+        - name: argo-server
+          args:
+            - server
+            - --namespaced
+            - --auth-mode
```
Comment (alexec) [inline manifests/quick-start/base/overlays/argo-server-deployment.yaml:13]: * [ ] Check to make sure `--auth-mode sso` does not result in both `server` and `sso`.

File: server/auth/gatekeeper.go [hunk 2]
```diff
@@ -21,24 +24,23 @@ type ContextKey string
 const (
 	WfKey   ContextKey = "versioned.Interface"
 	KubeKey ContextKey = "kubernetes.Interface"
-)
-
-const (
-	Client = "client"
-	Server = "server"
-	Hybrid = "hybrid"
+	UserKey ContextKey = "v1alpha1.User"
```
Comment (alexec) [inline server/auth/gatekeeper.go:27]: now we have the user, how about we mark each workflow with their creator

File: cmd/argo/commands/server.go [hunk 3]
```diff
@@ -104,7 +109,7 @@ See %s`, help.ArgoSever),
 		defaultBaseHRef = "/"
 	}
 	command.Flags().StringVar(&baseHRef, "basehref", defaultBaseHRef, "Value for base href in index.html. Used if the server is running behind reverse proxy under subpath different from /. Defaults to the environment variable BASE_HREF.")
-	command.Flags().StringVar(&authMode, "auth-mode", "server", "API server authentication mode. One of: client|server|hybrid")
+	command.Flags().StringArrayVar(&authModes, "auth-mode", []string{"server"}, "API server authentication mode. One of: client|server|hybrid")
```
Comment (jessesuen) [inline cmd/argo/commands/server.go:112]: Why is this changed to be repeatable? What is the behavior of `--auth-mode server --auth-mode hybrid` ?

Comment (jessesuen) [reply cmd/argo/commands/server.go:112]: I see it was to add `sso` as an option. I feel like the presence of SSO configuration should presume that option is available.

File: server/auth/oauth2/service.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+func (s *service) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	http.Redirect(w, r, s.config.AuthCodeURL("TODO"), http.StatusFound)
+}
+
+func (s *service) HandleCallback(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	state := r.URL.Query().Get("state")
+	if state != "TODO" {
```
Comment (alexec) [inline server/auth/oauth2/service.go:73]: TODO

File: server/auth/oauth2/service.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+		Endpoint:     provider.Endpoint(),
+		Scopes:       []string{oidc.ScopeOpenID, "groups"},
+	}
+	idTokenVerifier := provider.Verifier(&oidc.Config{ClientID: clientID})
+	log.WithFields(log.Fields{"redirectURL": config.RedirectURL, "issuer": issuer, "clientId": clientID}).Info("SSO configuration")
+	return &service{config, idTokenVerifier, baseHRef}, nil
+}
+
+func (s *service) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	http.Redirect(w, r, s.config.AuthCodeURL("TODO"), http.StatusFound)
```
Comment (alexec) [inline server/auth/oauth2/service.go:67]: TODO

File: server/auth/oauth2/service.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	ctx := r.Context()
+	state := r.URL.Query().Get("state")
+	cookie, err := r.Cookie(stateCookieName)
+	http.SetCookie(w, &http.Cookie{Name: stateCookieName, MaxAge: 0})
+	if err != nil {
+		w.WriteHeader(400)
+		_, _ = w.Write([]byte(fmt.Sprintf("invalid state: %v", err)))
+		return
+	}
+	if state != cookie.Value {
```
Comment (jessesuen) [inline server/auth/oauth2/service.go:94]: The state nonce is supposed to be stored on the server for a short period of time and verified by the **_server_** to be still valid upon return. Adding the state nonce as a cookie adds zero security as the cookie is verified against the query param, which the caller has control over both. 

It's better off removing state verification entirely and leave it as a TODO item. Better would be to use an in-memory time-limited cache of state nonces. However in-memory state nonces precludes more than 1 replica argo-server for SSO login, and then you would have to use something like redis.

File: server/auth/oauth2/service.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+		return wfv1.NullUser, fmt.Errorf("failed to verify id_token %v", err)
+	}
+	c := &claims{}
+	if err := idToken.Claims(c); err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to parse claims: %v", err)
+	}
+	return wfv1.User{Name: idToken.Subject, Groups: c.Groups}, nil
+}
+
+func (s *service) IsSSO(authorization string) bool {
```
Comment (jessesuen) [inline server/auth/oauth2/service.go:145]: I don't agree that this method should exist. A JWT is a JWT. It reasonable for a JWT to be the complete contents of the auth header. If you need to determine whether or not it is SSO (versus kubernetes token), this can be determined by the issuer of the JWT. Not through some invented auth token prefix..

File: docs/argo-server-rbac.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+> v2.9 and after
+
+## To start the server with RBAC:
+
+Firstly, you may wish to enable [SSO](argo-server-sso.md). The groups provided by your OAuth 2 provider and will be used as roles.
+
+Then, configure RBAC in the [config map](workflow-controller-configmap.yaml).
+
+Tip: use the [Casbin editor](https://casbin.org/editor/) to test out your policy. The model can be found in [rbac/service.go](../server/rbac/service.go).
```
[the remark this answers is not in the document]
Comment (alexec) [reply docs/argo-server-rbac.md:13]: ops - need to delete everything about RBAC

File: cmd/argo/commands/server.go [hunk 9]
```diff
@@ -104,7 +109,7 @@ See %s`, help.ArgoSever),
 		defaultBaseHRef = "/"
 	}
 	command.Flags().StringVar(&baseHRef, "basehref", defaultBaseHRef, "Value for base href in index.html. Used if the server is running behind reverse proxy under subpath different from /. Defaults to the environment variable BASE_HREF.")
-	command.Flags().StringVar(&authMode, "auth-mode", "server", "API server authentication mode. One of: client|server|hybrid")
+	command.Flags().StringArrayVar(&authModes, "auth-mode", []string{"server"}, "API server authentication mode. One of: client|server|hybrid|sso")
```
[the remark this answers is not in the document]
Comment (alexec) [reply cmd/argo/commands/server.go:112]: `hybrid` is converted into `server`,`client` - i.e. it is just sugar

File: server/auth/oauth2/service.go (hunk 6 above)
Comment (alexec) [reply server/auth/oauth2/service.go:94]: this was what I was thinking, I did a small amount of research:

https://security.stackexchange.com/questions/140883/is-it-safe-to-store-the-state-parameter-value-in-cookie

File: docs/argo-server-auth-mode.md [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# Auth
+
+You can choose which kube config the Argo Server uses:
+
+* "server" - in hosted mode, use the kube config of service account, in local mode, use your local kube config.
+* "client" - requires clients to provide their Kubernetes bearer token and use that.
+* "hybrid" - use the client token if provided, fallback to the server token if note.
+* "sso" - use single sign-on, this will use your Argo Server's service account as per "server".
```
[the remark this answers is not in the document]
Comment (alexec) [reply docs/argo-server-auth-mode.md:8]: correct - I would expect us to implement RBAC separately - it'll just use the argo server SA today

File: cmd/argo/commands/server.go [hunk 11]
```diff
@@ -35,14 +37,12 @@ func NewServerCommand() *cobra.Command {
 		Short: "Start the Argo Server",
 		Example: fmt.Sprintf(`
 See %s`, help.ArgoSever),
-		RunE: func(c *cobra.Command, args []string) error {
+		Run: func(c *cobra.Command, args []string) {
```
[the remark this answers is not in the document]
Comment (alexec) [reply cmd/argo/commands/server.go:40]: we are at some point - I think that @sarabala1979 has a plan

File: server/auth/oauth2/service.go (hunk 7 above)
Comment (alexec) [reply server/auth/oauth2/service.go:145]: Are you saying if (a) the token does not start with `Bearer` or `Basic` and (b) it is a JWET then assume it is SSO?

Comment (jessesuen) [reply server/auth/oauth2/service.go]: No, the k8s and the SSO tokens are both `Bearer` tokens, both in the form of a JWT. The problem, is how to distinguish between a JWT issued by K8s versus SSO. As I understand your implementation, you look at a specialized prefix. However, I am suggesting you can just look at the issuer, rather than rely on a specialized prefix.

I should rephrase that `IsSSO()` can exist, but it would be changed to unmarshal the JWT to switch on the issuer.

File: docs/argo-server-auth-mode.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jessesuen) [reply docs/argo-server-auth-mode.md:8]: Can you add a comment stating that it currently will use the server's ServiceAccount, but is expected change in the future to require explicit configuration.

File: server/auth/oauth2/service.go (hunk 7 above)
Comment (alexec) [reply server/auth/oauth2/service.go]: Ah. No, I think you maybe wrong on this. To clarify, we accept an "authorization" header. Today,  can be one of the following:

1. Empty string.
2. "Bearer ${bearerToken}" - i.e. K8S 
3. "Basic base64(${username}:${password)" - i.e. K8S

We need a way to determine what kind of token it is - but it is our token - it can whatever we want. I think bearer tokens are  opaque - we don't know if it is JWT or other - this is up to the Kubernetes provider.

Either way, it does not matter. The "authorization" header can be whatever we want, but we should be able to inspect it and make determination - we don't need to be smart about it - a dumb prefix will be simple and robust.

Hence this func:

https://github.com/argoproj/argo/pull/2745/files#diff-8ac7522dba311812e435bdd519739dddR35

Comment (alexec) [reply server/auth/oauth2/service.go]: As promised I revisited this and found a third-way. Authorization header can now be the following:

* `Bearer id_token:${idToken}`
* `Bearer ${k8sBearerToken}`
* `Basic ${k8sBasic}`

This is both simple (no complex and costly code to unpack token) and standardized (what @jessesuen wanted)

**Comment by alexec** [comment]
TestServer_GetWFClient/SSO

**Comment by alexec** [comment]
Assigning @jessesuen so I know who to chase :)

File: server/auth/sso/sso.go [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+	}
+	idToken, err := s.idTokenVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to verify id_token %v", err)
+	}
+	c := &claims{}
+	if err := idToken.Claims(c); err != nil {
+		return wfv1.NullUser, fmt.Errorf("failed to parse claims: %v", err)
+	}
+	return wfv1.User{Name: idToken.Subject, Groups: c.Groups}, nil
```
Comment (jessesuen) [inline server/auth/sso/sso.go:168]: I don't think we should have an API to return an invented `User` object. `groups` is not a required OIDC field and will not always be returned. Mapping `sub` to `name` is also not any standard. In our experience, there are many users who do not use the `groups` field and key off of a different or custom field for RBAC. For this to be future proof, we really we should return back the raw claims.

File: server/auth/sso/sso.go [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+const stateCookieName = "oauthState"
+
+func (s *sso) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	state := rand.RandString(10)
+	http.SetCookie(w, &http.Cookie{
+		Name:     stateCookieName,
+		Value:    state,
+		Expires:  time.Now().Add(1 * time.Hour),
```
Comment (jessesuen) [inline server/auth/sso/sso.go:88]: I believe state nonces are supposed to redeem on the order of minutes. I would set this much lower: 3 minutes.

**Comment by alexec** [comment]
@jessesuen I've 

* removed the user API completely - we don't need it for SSO

I've left the `user-info` page - it just redirects to the login page.
